src/core/or/circuitbuild.c

```diff
@@ -2418,17 +2414,29 @@ count_acceptable_nodes, (smartlist_t *nodes))
     if (! node->is_valid)
 //      log_debug(LD_CIRC,"Nope, the directory says %d is not valid.",i);
       continue;
-    if (! node_has_any_descriptor(node))
+    /* If we can use the node, check if we can connect directly. */
+    if (node_has_preferred_descriptor(node, 1))
+      ++direct;
+    else if (node_has_preferred_descriptor(node, 0))
+      ++indirect;
+    else
       continue;
     /* The node has a descriptor, so we can just check the ntor key directly */
     if (!node_has_curve25519_onion_key(node))
       continue;
-    ++num;
   } SMARTLIST_FOREACH_END(node);
 
 //    log_debug(LD_CIRC,"I like %d. num_acceptable_routers now %d.",i, num);
 
-  return num;
+  /* If the number of guard nodes (direct) is greater than one, and the
+   * total number of guard and non-guard nodes (indirect) is greater than
+   * 3, return the route length. Else, return 0 to signify that we do not
+   * have a complete circuit.
+   *
+   * Keep in mind that this function can succeed if direct is greater than
+   * 1 but indirect isn't because guard nodes can also be middle or exit
+   * nodes if configured so on the relay side. */
+  return (direct >= 1) && (direct + indirect >= 3) ? direct + indirect : 0;
```

Comment:
I think this calculation isn't what we actually want.

After all, we only use this function in one place: to tell if we have enough routers to build a path.  Maybe the right solution here is to have this function return the direct and indirect counts separately, and let new_route_len() decide whether it likes it?